### ISSUE TITLE
Attention模块

### ISSUE THREAD

**Issue by songqi-github** [SHI-Labs/Pyramid-Attention-Networks#12]
在PyramidAttention有如下代码：

wi = torch.cat([w[i][idx][0] for i in range(len(self.scale))], dim=0)  # [L, C, k, k]
问题： 取[i]后得到的shape为 [N, L, C, k, k], 进一步取 [idx] 得到[L, C, k, k]，最后的[0]是什么意义呢？ 另外，最后把所有scale的数据cat起来时，尺度不是不相同吗 怎么可以concat呢？而且cat之后的L是不是变成了5L（因为一共有5个scale）？
        
**yi = F.softmax(yi * self.softmax_scale, dim=1)**
if self.average == False:
      yi = (yi == yi.max(dim=1, keepdim=True)[0]).float()
raw_wi = torch.cat([raw_w[i][idx][0] for i in range(len(self.scale))], dim=0)
**yi = F.conv_transpose2d(yi, raw_wi, stride=self.stride, padding=1) / 4.**
y.append(yi)

问题：
1. 为什么在进行softmax时要乘以 self.softmax_scale呢？
2. 反卷积之后为什么要除以4？

**Comment by songqi-github** [comment]
if self.average == False:
    yi = (yi == yi.max(dim=1, keepdim=True)[0]).float()
这个代码的意义是什么呢？

**Comment by HarukiYqM** [comment]
> 在PyramidAttention有如下代码：
> 
> wi = torch.cat([w[i][idx][0] for i in range(len(self.scale))], dim=0) # [L, C, k, k]
> 问题： 取[i]后得到的shape为 [N, L, C, k, k], 进一步取 [idx] 得到[L, C, k, k]，最后的[0]是什么意义呢？ 另外，最后把所有scale的数据cat起来时，尺度不是不相同吗 怎么可以concat呢？而且cat之后的L是不是变成了5L（因为一共有5个scale）？
> 
> **yi = F.softmax(yi * self.softmax_scale, dim=1)**
> if self.average == False:
> yi = (yi == yi.max(dim=1, keepdim=True)[0]).float()
> raw_wi = torch.cat([raw_w[i][idx][0] for i in range(len(self.scale))], dim=0)
> **yi = F.conv_transpose2d(yi, raw_wi, stride=self.stride, padding=1) / 4.**
> y.append(yi)
> 
> 问题：
> 
> 1. 为什么在进行softmax时要乘以 self.softmax_scale呢？
> 2. 反卷积之后为什么要除以4？

1&2 是两个非核心的超参：
1. 参考inpainting方法https://arxiv.org/pdf/1801.07892.pdf
本文采用了跟该方法同样的数值，没有具体调整。
2. rescale feature map，避免值过大。其他合适的数值（4-6）均可达到相似的效果。

**Comment by HarukiYqM** [comment]
> if self.average == False:
> yi = (yi == yi.max(dim=1, keepdim=True)[0]).float()
> 这个代码的意义是什么呢？

这个在panet里一直为true。意义是只选取相似度的最高的patch参与反卷积 而忽略其他的。

**Comment by songqi-github** [comment]
wi = torch.cat([w[i][idx][0] for i in range(len(self.scale))], dim=0) # [L, C, k, k]
问题： 取[i]后得到的shape为 [N, L, C, k, k], 进一步取 [idx] 得到[L, C, k, k]，最后的[0]是什么意义呢？ 另外，最后把所有scale的数据cat起来时，尺度不是不相同吗 怎么可以concat呢？而且cat之后的L是不是变成了5L（因为一共有5个scale）？

**Comment by HarukiYqM** [comment]
> wi = torch.cat([w[i][idx][0] for i in range(len(self.scale))], dim=0) # [L, C, k, k]
> 问题： 取[i]后得到的shape为 [N, L, C, k, k], 进一步取 [idx] 得到[L, C, k, k]，最后的[0]是什么意义呢？ 另外，最后把所有scale的数据cat起来时，尺度不是不相同吗 怎么可以concat呢？而且cat之后的L是不是变成了5L（因为一共有5个scale）？

取[idx]后是[1,L,C,k,k]。[0]是为了去掉1。尺度是相同的，可以cat，意义是对所有scale做attention的dot product。是5L

**Comment by songqi-github** [comment]
懂了，谢谢您的解答
